nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -200,16 +227,19 @@ in
           assertion = (!(hasAttrByPath ["federation"] cfg.settings)) && (!(hasAttrByPath ["federation" "enabled"] cfg.settings));
           message = "`services.lemmy.settings.federation` was removed in 0.17.0 and no longer has any effect";
         }
+        {
+          assertion = cfg.database.uriFile != null -> cfg.database.uri == null && !cfg.database.createLocally && hasAttrByPath [ "database" "uri" "_secret" ] cfg.settings;
```

Comment:
Do we really need these assertions? What about other secret options? Is it enough to mention that the `fooFile` option will override the `foo` option in the description of `fooFile`?

Reply:
Normally they are not needed since setting secret files causes `{ _secret = ... }`  to be assigned to the setting which causes a `is defined multiple times` error if the user assigned something different.
`database.uri` is different since it does not set `settings.database.uri` which means it is possible to to assign different things.
The `hasAttrByPath` check can be removed.

Reply:
Now I see that `database.uri` is special because `LEMMY_DATABASE_URL` will be set if `database.uri` is set.  I guess if `LEMMY_DATABASE_URL` is not set, database uri will be read from the config file.

> causes a is defined multiple times error

No error will be raised because they are merged by `recursiveUpdate`.

> database.uri is different since it does not set settings.database.uri

I am a bit confused.  What do you mean by "it does not set settings.database.uri"?

Reply:
> No error will be raised because they are merged by recursiveUpdate.

`recursiveUpdate` only merges the defaults and secrets, which are then assigned to `services.lemmy.settings`.
If a user sets `services.lemmy.settings.setup.admin_password` and `services.lemmy.adminPasswordFile` it will cause an error since both assign something different to `services.lemmy.settings.setup.admin_password`.

> I am a bit confused. What do you mean by "it does not set settings.database.uri"?

Setting `database.uri` to `...` does not set `settings.database.uri` to `...`.